### DISCUSSION TITLE
想请问一下微软账户的授权是怎么刷新的？

### DISCUSSION THREAD

**Discussion by BrandonStudio** [spencerwooo/onedrive-vercel-index#666] [Q&A] [answered]
账户权限似乎最多只有两年，这个项目是定时刷新授权的吗？可是好像没有用到相关权限呀

**Comment by BrandonStudio** [comment]
我再补充一下，整个程序里相当于有三个令牌，一个是应用程序本身的令牌，即 Client ID 和 Client Secret 组合，这个令牌的最长时限是两年，而且似乎是写死在程序里的，这个是怎么刷新的，或者根本不刷新？如果使用证书替代 Client Secret，是否就没有时限了？
第二个和第三个是部署后存入 Upstash 的 AccessToken 和 RefreshToken，AccessToken 的有效期似乎是1个小时，RefreshToken 的有效期似乎是90天，这两个令牌是何时如何刷新的？

**Reply by githubhjs** [reply]
同問... multi-factor authentication 啟用之後...重啟不能

**Comment by myl7** [comment, marked as answer]
> Client ID 和 Client Secret

这个需要对应 client 的创建者手动改代码进行刷新

> 证书替代 Client Secret

这个没了解过不太清除，client id + secret 应该是比较普遍的用法（例如 rclone 等也是这样的）

> AccessToken 和 RefreshToken
>
> 这两个令牌是何时如何刷新的？

access token 通过携带 refresh token 访问特定 API 进行刷新，在此过程中如果 refresh token 也需要刷新将会在 access token 刷新的 response 内带上刷新好的 refresh token，此部分 OneDrive 文档有具体介绍

**Reply by BrandonStudio** [reply]
就是说 refresh token 刷新之后，90天有效期也会刷新对吧？
我看文档的时候觉得是不管怎么刷新 refresh token 的到期时间距离首次申请都是90天[]

**Reply by myl7** [reply]
> 就是说 refresh token 刷新之后，90天有效期也会刷新对吧？

是的，本项目此部分代码的添加过程可以参考 #187
